Migrate updateQuestion to MLv2

ranquild opened the pull request:
Closes https://github.com/metabase/metabase/issues/37392

Simply the QB logic by removing `hasNewColumns` check. Now we decide that we need to-rerun the query in place where we change the query itself and not try to figure it out later.

ranquild committed:
Migrate updateQuestion to MLv2

uladzimirdev commented on frontend/src/metabase/query_builder/actions/core/updateQuestion.ts:
```diff
@@ -29,15 +29,23 @@ import { onCloseQuestionInfo, setQueryBuilderMode } from "../ui";
+  if (!queryResult) {
+    return false;
+  }
+  if (question.isNative()) {
```
this change is not clear to me

kamilmielnik replied:
It's due to `query instanceof StructuredQuery` part.

ranquild committed:
Fix column check

ranquild committed:
Merge master

ranquild committed:
Workaround incorrect column names

ranquild committed:
Workaround incorrect column names

kamilmielnik commented on frontend/src/metabase/query_builder/actions/core/updateQuestion.ts:
```diff
@@ -29,15 +29,22 @@ import { onCloseQuestionInfo, setQueryBuilderMode } from "../ui";
+  const query = question.query();
+  const stageIndex = -1;
+  const queryColumns = Lib.returnedColumns(query, stageIndex);
+  return queryColumns.length > queryResultColumns.length;
```
I think this is not enough. Previously diff of arrays of names of columns was checked.
So it does not handle a case where 1 column is removed and another is added.

uladzimirdev replied:
nice catch btw

ranquild replied:
I'm experimenting to see if I can avoid the issue with the wrong column names in 
`returnedColumns`. But it seems that not. Reverted this change. The PR would have to stay here until the referenced MLv2 bug is fixed

ranquild committed:
Workaround incorrect column names

metamben committed:
Introduce same-column-lists?

Fixes #37517.

This function makes it possible for the FE to decide if the query has to be re-run.

ranquild committed:
Merge branch 'master' into 37517-returned-columns-duplicate-names

ranquild committed:
Merge remote-tracking branch 'origin/37517-returned-columns-duplicate-names' into update-question-legacy-query

ranquild committed:
Use isSameColumnList

ranquild committed:
Pass "run" instead of checking for columns

ranquild committed:
Remove CLJS changes

ranquild committed:
Fix tests

ranquild commented on e2e/test/scenarios/visualizations-charts/reproductions/21452-xhr-on-every-char-for-rename.cy.spec.js:
```diff
@@ -36,8 +36,6 @@ describe("issue 21452", () => {
     // eslint-disable-next-line no-unscoped-text-selectors -- deprecated usage
     cy.findByText("Display type").click();
-    // Blur will result in another POST request which is expected
-    cy.wait("@dataset");
```
New streamline logic works even before than before - there is no duplicate request when changing viz settings only.

ranquild commented:
@kamilmielnik could you take a look one more time, please? The PR has significantly changed.

kamilmielnik reviewed (approved):
Nice :+1:
